File: 2022/python_2022/day_03.py
def common_char_finder(words: list) -> str:
    word_set = set(words[0])
    for word in words[1:]:
        word_set = word_set.intersection(set(word))

    return "".join(word_set)


def char_to_priority(char: str) -> int:
    if char.islower():
        return ord(char) - 96
    else:
        return ord(char) - 38


# Part 1
file = open("../inputs/day_03.txt", "r")

priorities = []
for line in file:
    common_char = "".join(
        common_char_finder(
            [line.strip()[: len(line) // 2], line.strip()[len(line) // 2 :]]
        )
    )
    priorities.append(char_to_priority(common_char))

print(f"Part 1: {sum(priorities)}")

# Part 2
file = open("../inputs/day_03.txt", "r")
priorities = []

# Check common chars by groups of 3 lines
lines = file.readlines()
for i in range(0, len(lines), 3):
    common_char = "".join(
        common_char_finder(
            [
                lines[i].strip(),
                lines[i + 1].strip(),
                lines[i + 2].strip(),
            ]
        )
    )
    priorities.append(char_to_priority(common_char))

print(f"Part 2: {sum(priorities)}")

File: 2022/rust_2022/src/day_03.rs
use std::{collections::HashSet, fs};

fn common_char_finder(words: Vec<&str>) -> char {
    let mut common_chars = HashSet::<char>::new();
    common_chars.extend(words[0].chars());

    for word in words.iter().skip(1) {
        let word_set = word.chars().collect::<HashSet<char>>();
        common_chars = common_chars
            .intersection(&word_set)
            .map(|c| *c)
            .collect::<HashSet<char>>();
    }

    // Guaranteed to always be a single char
    *common_chars.iter().next().unwrap()
}

fn char_to_priority(char: char) -> i32 {
    if char.is_lowercase() {
        char as i32 - 96
    } else {
        char as i32 - 38
    }
}

pub fn part_1() -> i32 {
    fs::read_to_string("../inputs/day_03.txt")
        .unwrap()
        .split("\n")
        .map(|line| {
            let common_char =
                common_char_finder(vec![&line[0..line.len() / 2], &line[line.len() / 2..]]);
            char_to_priority(common_char)
        })
        .sum()
}

pub fn part_2() -> i32 {
    fs::read_to_string("../inputs/day_03.txt")
        .unwrap()
        .split("\n")
        .collect::<Vec<&str>>()
        .chunks(3)
        .map(|chunk| {
            let common_char = common_char_finder(vec![&chunk[0], &chunk[1], &chunk[2]]);

            char_to_priority(common_char)
        })
        .sum()
}
